exporter/awskinesisexporter/internal/batch/batch.go

```diff
@@ -0,0 +1,130 @@
+	"go.opentelemetry.io/collector/consumer/consumererror"
+	protov2 "google.golang.org/protobuf/proto"
+)
+
+const (
+	MaxRecordSize     = 1 << 20 // 1MiB
+	MaxBatchedRecords = 500
+)
+
+var (
+	// ErrPartitionKeyLength is used when the given key exceeds the allowed kinesis limit of 256 characters
+	ErrPartitionKeyLength = errors.New("partition key size is greater than 256 characters")
+	// ErrRecordLength is used when attempted record results in a byte array greater than 1MiB
+	ErrRecordLength = consumererror.Permanent(errors.New("record size is greater than 1 MiB"))
+)
+
+type Batch struct {
+	maxBatchSize  int
+	maxRecordSize int
+
+	records []*kinesis.PutRecordsRequestEntry
+}
+
+type Option func(bt *Batch)
+
+func WithMaxRecordsPerBatch(limit int) Option {
+	return func(bt *Batch) {
+		if MaxBatchedRecords < limit {
+			limit = MaxBatchedRecords
+		}
+		bt.maxBatchSize = limit
+	}
+}
+
+func WithMaxRecordSize(size int) Option {
+	return func(bt *Batch) {
+		if MaxRecordSize < size {
+			size = MaxRecordSize
+		}
+		bt.maxRecordSize = size
+	}
+}
+
+func New(opts ...Option) *Batch {
+	bt := &Batch{
+		maxBatchSize:  MaxBatchedRecords,
+		maxRecordSize: MaxRecordSize,
+		records:       make([]*kinesis.PutRecordsRequestEntry, 0, MaxRecordSize),
+	}
+
+	for _, op := range opts {
+		op(bt)
+	}
+
+	return bt
+}
+
+func (b *Batch) addRaw(raw []byte, key string) error {
+	if l := len(key); l == 0 || l > 256 {
+		return ErrPartitionKeyLength
+	}
+
+	if len(raw) > b.maxRecordSize {
+		return ErrRecordLength
+	}
+
+	b.records = append(b.records, &kinesis.PutRecordsRequestEntry{Data: raw, PartitionKey: aws.String(key)})
+	return nil
+}
+
+// AddProtobufV1 allows for deprecated protobuf generated types to be exported
+// and marshaled into records to be exported to kinesis.
+// The protobuf v1 message is considered deprecated so where possible to use
+// batch.AddProtobufV2.
+func (b *Batch) AddProtobufV1(message protov1.Message, key string) error {
+	data, err := protov1.Marshal(message)
+	if err != nil {
+		return err
+	}
+	return b.addRaw(data, key)
+}
+
+// AddProtobufV2 accepts the protobuf message and marshal it into a record
+// ready to submit to kinesis.
+func (b *Batch) AddProtobufV2(message protov2.Message, key string) error {
+	data, err := protov2.Marshal(message)
+	if err != nil {
+		return err
+	}
+	return b.addRaw(data, key)
+}
+
+// Chunk breaks up the iternal queue into blocks that can be used
+// to be written to he kinesis.PutRecords endpoint
+func (b *Batch) Chunk() [][]*kinesis.PutRecordsRequestEntry {
+	length, chunk := len(b.records), make([][]*kinesis.PutRecordsRequestEntry, 0, len(b.records)/b.maxBatchSize+1)
+	for i, end := 0, min(length, b.maxBatchSize); i < length; i, end = i+b.maxBatchSize, i+end+min(b.maxBatchSize, length-end) {
+		chunk = append(chunk, b.records[i:end])
+	}
+	return chunk
+}
+
+func min(a, b int) int {
```

Comment:
I can't tell you how much i hate having this here and eagerly await the promise of generics in go 1.18